Add CIP for weighted validator liveness rewards (#3)
moritzkiefer-da/cips: cip-00XX/cip-XXXX.md

supporting weights defined through governance.

This allows multi-hosting of validators while still providing the same
rewards to the parties on the same node as if they were running their
own node. It also allows weighting certain some parties higher, e.g.,
because they have been shown to contribute significantly to the
success of the network.

## Copyright

This CIP is licensed under CC0-1.0: [Creative Commons CC0 1.0 Universal](https://creativecommons.org/publicdomain/zero/1.0/)

Note: other licenses are avilable see recommendations in [cip-0000](../../cips/cip-0000/cip-0000.md)

## Specification

The existing functionality where each SV can unilaterally grant a
`ValidatorLicense` to any validator operator by providing a validator
onboarding secret, is extended such that each SV can unilaterally
grant a `ValidatorLicense` with the default weight to arbitrary
parties (including parties relying on external signing) through the SV
UI.

In addition to that, SVs get the ability to change the weight of a
validator license to a non-negative integer. The liveness rewards
minted by a license with weight N are equivalent to the liveness

> rewards minted by N licenses.

Reviewer: What's the motivation for the integer? Why not use the flexibility we get from a Decimal?

Author: Good question. I intuitively copied what we have for SV weights but I think you're right a Decimal is better.

### SV UI

The SV UI is extended with three different functionalities:

1. An SV can grant a validator license to a user-specified party id.
2. SVs can vote on weight changes for an existing validator license.
3. SVs can vote on withdrawing a validator license.

### Daml

- At the Daml level there is no distinction between validator operator
  parties and any other parties. So the existing choices that allow an
  SV to grant a validator license to validator operator parties can
  already be used to grant licenses to arbitrary parties.

> - To support parties relying on external signing, a new

Reviewer: ```suggestion
- To support automatic reward minting for parties relying on external signing, a new
```

  `MintingDelegation` contract is added which allows signing an
  transaction once which delegates the right to create and mint
  liveness activity records to a validator operator party. This then
  allows the automation in the validator app to submit the required
  transactions each round on behalf of the external party. Note that
  `MintingDelegation` is limited to creating and minting activity
  records, it does not grant the delegate rights to do anything else

>   on behalf of the party.

Reviewer: How are these `MintingDelegation` contracts accepted and renewed? UI? Config? API call?

- To support weighting of liveness rewards, the `ValidatorLicense` and
  `ValidatorLivenessActivityRecord` are extended with an optional
  weight field. If not set, as is the case for all existing licenses,
  it defaults to weight 1.

## Motivation

MainNet is close to the maximum number of validators that can be
supported. This is driven to a significant degree due to spinning up a
separate validator being the only option to get more liveness rewards.
By supporting liveness rewards for non-validator operator parties,
this pressure is reduced without impacting the rewards that users can
receive.

Weighting allows balancing the benefit to the network coming from

> users that do nothing than accumulate validator rewards with parties

Reviewer: ```suggestion
users that do nothing other than accumulating validator rewards with parties
```

that significantly contribute to the success of the network.

## Rationale

Manual weighting through votes provides a simple mechanism to balance
the contributions of different parties to the network. While more
complex automatic mechanisms based on staking, measuring activity in
apps or similar could be built, those all have vastly more

> implementation complexity and could be added on top of this by

Reviewer: ```suggestion
implementation complexity and could be added on top of this cIP by
```

automating votes on weight changes.

Allowing SVs to unilaterally grant a license with default weight to
any party is already supported by the current Daml code and consistent
with being able to unilaterally grant it to validator operators. If
this gets abused, other SV operators can withdraw validator licenses

> or set the weight to 0.

Reviewer: ```suggestion
or set their weight to 0.
```

Weight changes on the other hand should be made based on governance

> decisions so those require a vote.

Reviewer: ```suggestion
decisions, so those require a vote.
```

## Backwards compatibility

> The changes are all backwards compatible. However, as any Daml change

Reviewer: ```suggestion
The changes are all backwards compatible. However, like any Daml change
```

## Weighted Validator Liveness Rewards for SV-Determined Parties

<pre>
  CIP: ?
* Layer: Daml

>   Title: Weighted Validator Liveness Rewards for SV Chosen Parties

Reviewer 2: ```suggestion
  Title: Weighted Validator Liveness Rewards for SV-Chosen Parties
```

  Author: Moritz Kiefer
* Discussions-To: cip-discuss@list.sync.global
* Comments-Summary: N/A
* Comments-URI: N/A
  Status: Draft
  Type: Tokenomics
  Created: 2025-07-23
  License: CC0-1.0
* Post-History: N/A
</pre>

## Abstract

This CIP proposes extending the concept of validator liveness rewards

> to arbitrary parties not just validator operators as well as

Reviewer 2: ```suggestion
to arbitrary parties--not just validator operators--as well as
```

supporting weights defined through governance.

> This allows multi-hosting of validators while still providing the same

Reviewer 2: ```suggestion
This allows hosting multiple parties on a single validator while still providing the same
```

rewards to the parties on the same node as if they were running their
own node. It also allows weighting certain some parties higher, e.g.,

> because they have been shown to contribute significantly to the

Reviewer 2: ```suggestion
because they have been shown, or have committed, to contribute significantly to the
```

> success of the network.

Reviewer 2: ```suggestion
success of the network, or have committed to do so. 
```

## Copyright

This CIP is licensed under CC0-1.0: [Creative Commons CC0 1.0 Universal](https://creativecommons.org/publicdomain/zero/1.0/)

Note: other licenses are avilable see recommendations in [cip-0000](../../cips/cip-0000/cip-0000.md)

## Specification

The existing functionality where each SV can grant a
`ValidatorLicense` to any validator operator by providing a validator
onboarding secret, is extended such that each SV can
grant a `ValidatorLicense` with the default weight to arbitrary
parties (including parties relying on external signing) through the SV

> UI. This expected to only be done after approval from the tokenomics commitee.

Reviewer 2: ```suggestion
UI. 
```

Changed my mind; probably best to just say the SV can do this, and leave governance to offline processes.

  party. Note that `MintingDelegation` is limited to creating and
  minting activity records, it does not grant the delegate rights to
  do anything else on behalf of the party. Minting delegations can be
  created and accepted through the splice wallet UI for local parties
  or through the ledger API endpoints for [external transaction
  signing](https://docs.digitalasset.com/build/3.3/tutorials/app-dev/external_signing_submission.html)
  for parties relying on external signing.
- To support weighting of liveness rewards, the `ValidatorLicense` and
  `ValidatorLivenessActivityRecord` are extended with an optional
  weight field. If not set, as is the case for all existing licenses,
  it defaults to weight 1.

## Motivation

MainNet is close to the maximum number of validators that can be
supported. This is driven to a significant degree due to spinning up a
separate validator being the only option to get more liveness rewards.
By supporting liveness rewards for non-validator operator parties,
this pressure is reduced without impacting the rewards that users can
receive.

Weighting allows balancing the benefit to the network coming from

> users that do nothing other than accumulating validator rewards with parties

Reviewer 2: ```suggestion
users that do nothing other than accumulating validator rewards with the benefits provided by parties
```

that significantly contribute to the success of the network.

## Rationale

Manual weighting through votes provides a simple mechanism to balance
the contributions of different parties to the network. While more
complex automatic mechanisms based on staking, measuring activity in
apps or similar could be built, those all have vastly more
implementation complexity and could be added on top of this CIP by
automating votes on weight changes.

Allowing SVs to unilaterally grant a license with default weight to
any party is already supported by the current Daml code and consistent

> with being able to unilaterally grant it to validator operators. If

Reviewer 2: At least right now, this is not a unilateral right: the Tokenomics WG has to approve it. 

I think we should explicitly say that we're moving this right to add a Valdiator License / Liveness License from the GSF Tokenomics WG back to the SVs.

Author: I think there are two levels here:

1. At the Daml level it is a unilateral right now. That part doesn't change.
2. Currently, we expect a non-malicious SV to only do this after tokenomics approval. We could remove this requirement or keep it. Question for you and Eric I guess.

Author: I've rephrased it such that tokenomics approval is required like it is atm but not enforced through the Daml code as is the case for validator licenses atm. If that's not what we want let me know. Technically there is really no difference in here, just a matter of when we count an SV as malcious.

Author: So in particular, we might also consider two steps:

1. Keep this CIP with tokenomics approval. That may make it easier to get it approved as it is less of a change to procedures.
2. If desired that requirement could be lifted without any technical change.
